python/cudf/cudf/core/frame.py

```diff
@@ -1819,6 +1827,271 @@ def sample(
+
+        if (
+            cudf_non_category_frames.shape[1] > 0
+            and cudf_category_frames.shape[1] > 0
+        ):
+            result = cudf.concat(
+                [cudf_non_category_frames, cudf_category_frames], axis=1
+            )
+        elif cudf_non_category_frames.shape[1] > 0:
+            result = cudf_non_category_frames
+        else:
+            result = cudf_category_frames
+
+        if dtypes:
+            for name in result._data.names:
+                dtype = dtypes[name]
+                if dtype == "categorical":
+                    dtype = "category"
+                elif dtype == "date" or dtype == "datetime":
+                    dtype = "datetime64[ms]"
+                elif dtype == "empty":
+                    dtype = "object"
+
+                result._data[name] = result._data[name].astype(dtype)
+
+        return result[column_names]
+
+    @annotate("TO_ARROW", color="orange", domain="cudf_python")
+    def to_arrow(self):
+        """
+        Convert to arrow Table
+
+        Examples
+        --------
+        >>> import cudf
+        >>> df = cudf.DataFrame(
+        ...     {"a":[1, 2, 3], "b":[4, 5, 6]}, index=[1, 2, 3])
+        >>> df.to_arrow()
+        pyarrow.Table
+        a: int64
+        b: int64
+        index: int64
+        """
+
+        data = self.copy(deep=False)
+
+        codes = {}
+        # saving the name as they might get changed from int to str
+        codes_keys = []
+        categories = {}
+        # saving the name as they might get changed from int to str
+        names = self._data.names
+        null_arrays_names = []
+
+        for name in self._data.names:
+            col = self._data[name]
+            if isinstance(col, cudf.core.column.CategoricalColumn,):
+                # arrow doesn't support unsigned codes
+                signed_type = (
+                    min_signed_type(col.codes.max())
+                    if col.codes.size > 0
+                    else np.int8
+                )
+                codes[str(name)] = col.codes.astype(signed_type)
+                categories[str(name)] = col.categories
+                codes_keys.append(name)
+            elif isinstance(
+                col, cudf.core.column.StringColumn
+            ) and col.null_count == len(col):
+                null_arrays_names.append(name)
+
+        data = cudf.DataFrame(data._data)
+        data = data.drop(codes_keys + null_arrays_names, inplace=True)
+
+        out_table = pa.table([])
+        if data.shape[1] > 0:
+            out_table = libcudf.interop.to_arrow(
+                data, data._data.names, keep_index=False
+            )
+
+        # If the dicitonary array is a string array and of length `0`
+        # it should be a null array
+        def get_dictionary_array(array):
+            if isinstance(array, pa.StringArray) and len(array) == 0:
+                return pa.array([], type=pa.null())
+            else:
+                return array
```

Comment:
Make this an underscore method and move it out of the `to_arrow` definition.

Reply:
But this is not being used anywhere, so it would be waste to keep it anywhere else.

Reply:
Every time `to_arrow` is called this gets unnecessarily redefined currently. It's okay if it's only used in this function.